AmplifyPlugins/Core/AWSPluginsCore/Model/Support/Model+GraphQL.swift

```diff
@@ -18,76 +18,129 @@ extension Model {
                     } catch {
                         preconditionFailure("Could not turn into json object from \(value)")
                     }
                 }
+            case .string:
+                // The input may contain the same keys from `.model` case when the `getFieldNameForAssociatedModels`
+                // returns the same value as the `modelField.name`. If this is the case, let the `.model` case take
+                // precedent over the explicit string field on the Model by ignoring the value that is about to be added
+                if !input.keys.contains(modelField.name) {
+                    input[modelField.name] = value
+                }
             default:
-                input[name] = value
+                input[modelField.name] = value
             }
         }
+
+        return fixHasOneAssociationsWithExplicitFieldOnModel(input, modelSchema: modelSchema)
+    }
+
+    /// This is to account for Models with an explicit field on the Model along with an object representing the hasOne
+    /// association to another model. See https://github.com/aws-amplify/amplify-ios/issues/920 for more details.
+    /// When the associated object is `nil`, remove the key from the GraphQL input to prevent runtime failures.
+    /// When the associated object is found, take this value over the explicit field's value by replacing the correct
+    /// entry for the field name of the associated model.
+    private func fixHasOneAssociationsWithExplicitFieldOnModel(_ input: GraphQLInput,
+                                                               modelSchema: ModelSchema) -> GraphQLInput {
+        var input = input
+        modelSchema.fields.forEach {
+            let modelField = $0.value
+            if case .model = modelField.type,
+               case .hasOne = modelField.association,
+               input.keys.contains(modelField.name) {
+
+                let modelIdOrNilOptional = input[modelField.name]
+                // swiftlint:disable:next syntactic_sugar
+                guard case .some(Optional<Any>.some(let modelIdOrNil)) = modelIdOrNilOptional else {
+                    input.removeValue(forKey: modelField.name)
+                    return
+                }
+                if let modelIdValue = modelIdOrNil as? String {
+                    let fieldName = getFieldNameForAssociatedModels(modelField: modelField)
+                    input[fieldName] = modelIdValue
+                }
+            }
+        }
+
         return input
     }
+
+    private func getModelId(from value: Any, modelSchema: ModelSchema) -> String {
+        if let modelValue = value as? Model {
+            return modelValue.id
+        } else if let value = value as? [String: JSONValue],
+                  case .string(let primaryKeyValue) = value[modelSchema.primaryKey.name] {
+            return primaryKeyValue
+        }
+
+        preconditionFailure("Could not retrieve the identifier from associated model: \(value)")
```

Comment:
Hope this wont break? Previous code returned nil here.

Reply:
i wrote this on the assumption that it should always get an id from the value, but i'm more inclined to return nil here to introduce less changes to the code